### ISSUE TITLE
subscription render infinite loop with inline query 

### ISSUE THREAD

**Issue by alidcastano** [trojanowski/react-apollo-hooks#160]
not sure if this is a bug, or just improper usage, but either way I wanted to open up an issue since it took me a bit to figure out 

with an inline query (it's declared inside the function itself, see below) the subscription results in an infinite render

```
const { data } = useSubscription(gql`
  subscription  SomethingChanged {
     somethingChanged {
      id 
    }
  }
`)
```

the reason for this is that part of the inputs passed as the second param to `useEffect` is the `query` variable, and if it's declared inline (which perhaps it shouldn't be, but still), it'll be a new declaration in each render. 

thus, `useEffect` gets called, and since the cleanup function of `useSubscription` also forces a rerender (i.e. it calls ` setResult({ loading: true });`) it results in an infinite loop 

i didn't plan to leave the query declaration inline - was just testing it out in development and this bug took a bit of time to discover (i though I was doing something wrong in the server at first)

**Comment by DoMiNeLa10** [comment]
`gql` compiles the query to an Object used by apollo, so every time you call it, you create a new Object, which will be different from the previous value, so it will trigger the query to fire again. You'll need to move the query somewhere else, for instance into a variable outside of the component.

**Comment by alidcastano** [comment]
yeah but it can check if the query itself is the same (the inline syntax works with relay for example)

it's unintuitive that the same query would cause a rerender everytime, and it's much easier to test things would inline

**Comment by DoMiNeLa10** [comment]
You could set up your build system to compile queries at build time, which would get rid of the problem. I don't recommend that, as compiled queries can get pretty big, and having a couple of them can easily outweigh the cost of the query compiler itself.

Since the `gql` tag is just a function, you can only defer what it's doing at build time, otherwise calling it will cause it to run again and create a new object.

**Comment by alidcastano** [comment]
for context i'm totally fine with declaring the `gql` tag outside the function as its own variable

but while testing out a query this caught me off guard, so I figured I'd open up an issue about it

**Comment by DominicTobias-b1** [comment]
The only way I managed to solve the infinite loop issue so far is to do a shallow equal on the subscription object coming in via an array from the server and memoize :/ This at least breaks the infinite loop after 1 iteration. This is really nasty though...

```js
const emptyBids: OrderBookOrder[] = []
const emptyAsks: OrderBookOrder[] = []

function calcOrderbook(payload: OrderBook_datastore_order_book_funct) {
  let bids: OrderBookOrder[] = []
  let asks: OrderBookOrder[] = []

  if (payload.bids?.length) {
    bids = payload.bids.map((bid: OrderbookDBRecord) => ({
      id: `${bid.price}`,
      size: bid.price_level_qty,
      total: bid.price_level_qty * bid.price,
      price: bid.price,
    }))
  }

  if (payload.asks?.length) {
    asks = payload.asks.map((ask: OrderbookDBRecord) => ({
      id: `${ask.price}`,
      size: ask.price_level_qty,
      total: ask.price_level_qty * ask.price,
      price: ask.price,
    }))
  }

  return { bids, asks, lastPrice: payload.last_price }
}

const memoizedCalcOrderbook = memoizeOne(calcOrderbook, isEqual)

export const useOrderbookSubscription = (): OrderbookSubscription => {
  const { data, error } = useSubscription<OrderBook>(ORDERBOOK_SUBSCRIPTION)

  const formattedData = useMemo<OrderbookData>(() => {
    if (!data?.datastore_order_book_funct[0]) {
      return { bids: emptyBids, asks: emptyAsks }
    }

    const payload = data.datastore_order_book_funct[0]
    return memoizedCalcOrderbook(payload)
  }, [data])

  return { data: formattedData, error }
}
```

Otherwise the subscription causes the component to re-render, and Apollo re-subscribes to the endpoint and gets the same data again (in a loop) :|
